### ISSUE TITLE
bug DiffComparator when using the Thymeleaf if tag

### ISSUE THREAD

**Issue by dirkdeyne** [medusa-ui/medusa#358]
Branch [bug DiffComparator remove](https://github.com/medusa-ui/medusa/tree/rewrite-1.0.0-bug-DiffComparator-remove)

_'Do removals last'_ does not fix the problem when using Thymeleaf if-attribute.
Adding some extra `th:if` tags after the button messes things up again...

**Comment by dirkdeyne** [comment]
Branch [rewrite-1.0.0-wrap-th-if-tag](https://github.com/medusa-ui/medusa/tree/rewrite-1.0.0-wrap-th-if-tag) fixes bug by wrapping element with a `th:if` in a div. Elements like `tr` , `li`,... don't need to be wrapped.

**Comment by dirkdeyne** [comment]
Nested `th:if`-tag wrapped in  a `div` can get us in trouble. 
ref: [medusa-ui/src/test/java/io/getmedusa/medusa/core/render/WrapThIfTagsTest.java](https://github.com/medusa-ui/medusa/commit/8db889101d21d19ba47fb8e0c05857f2150fb449#diff-359f40fd2af141fa924c18542613e37d90454baa65817920e5cf060f9f75cf43)

**Comment by kevindeyne** [comment]
I have merged in the new diff engine, which hopefully fixes a lot of these types of problems.

**Comment by kevindeyne** [comment]
Added some more error scenarios to the diff engine's unit test

**Comment by dirkdeyne** [comment]
Detected where `ADDITION` go wrong:

We compare 'old' HTMLLayers with the 'new'  HTMLLayers.
XPath is used to identify an `HTMLLayer`.
An insertion `Delta` is created if the `HTMLLayer` does not exist in the 'old' HTML.
But what if the 'old' HTML already contains an element with that given XPath?
Then an insertionDelta is created but based on the wrong `HTMLLayer`

An example:
old
```html
<section>
      <p>1</p>
      <span>2</span>
      <p>3</p>             <!-- xpath : /body[1]/section[1]/p[2]  -->
</section>
```
new
```html
<section>
      <p>1</p>
      <p>4</p>            <!-- xpath : /body[1]/section[1]/p[2]  ignored because it also exists in old -->
      <span>2</span>
      <p>3</p>            <!-- xpath : /body[1]/section[1]/p[3]  wrongly detected as new -->
</section>
```
The second `p`-tag  (`/body[1]/section[1]/p[2]`) is new but is ignored because it also exists in 'old' 
The last `p`-tag exists in old but is marked as new because xpath(`/body[1]/section[1]/p[3]`) does not exist in old.

The generated result is wrong
```html
<section>
      <p>1</p>
      <span>2</span>
      <p>4</p>
      <p>3</p>   
</section>
```

**Comment by kevindeyne** [comment]
Oh ok, that makes a lot of sense. I forgot to account for the actual node position, then. Easy to fix.

Yes, in an ideal scenario you would know that the only change is that a span has moved up, but our layer approach is (intentionally) more naive to keep it simpler.

I will make a new version, let's see if this fixes things in 0.0.2 of the diff engine

**Comment by dirkdeyne** [comment]
Updated `medusa-ui` with `diff-engine 0.0.2` and the _complex_  [th-if sample page](http://localhost:8080/th-if) works fine!
So it looks like this bug can be closed.
